### PR TITLE
Allow UDP for AWS NLB

### LINKED ISSUE #79523: Support UDP with AWS NLB
**What happened**:
```Error creating load balancer (will retry): failed to ensure load balancer for service default/test: Only TCP LoadBalancer is supported for AWS ELB```

**What you expected to happen**: kubernetes creating an AWS NLB for UDP 

**How to reproduce it (as minimally and precisely as possible)**: Create service with annotation `service.beta.kubernetes.io/aws-load-balancer-type: nlb` and port with `protocol: UDP`

**Anything else we need to know?**:
AWS recently announced that UDP is available within their NLB https://aws.amazon.com/de/blogs/aws/new-udp-load-balancing-for-network-load-balancer/

**Environment**:
- Kubernetes version (use `kubectl version`): 1.11.9
- Cloud provider or hardware configuration: AWS

### PR DESCRIPTION
**Pull request by owenthereal**
**What type of PR is this?**

/kind bug
/sig aws

**What this PR does / why we need it**:

There is [check](https://github.com/kubernetes/cloud-provider-aws/blob/8f6e1e3d2cfa20a0deac088b9c4022d433500369/pkg/cloudprovider/providers/aws/aws.go#L3284-L3286) on protocol that doesn't consider NLB. For example, the followings

```yaml
kind: Service
apiVersion: v1
metadata:
  name: nlb
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
spec:
  ports:
    - name: port
      port: 443
      targetPort: 443
      protocol: UDP
  type: LoadBalancer
```

yield the error:

```
Warning  CreatingLoadBalancerFailed  0s (x2 over 5s)  service-controller  Error creating load balancer (will retry): failed to ensure load balancer for service default/release-name-openvpn: Only TCP LoadBalancer is supported for AWS ELB
```

**Which issue(s) this PR fixes**:

The changes ignore the check if it's an NLB with backing unit tests.

Fix https://github.com/kubernetes/kubernetes/issues/79523.

**Special notes for your reviewer**:

N/A

**Does this PR introduce a user-facing change?**:

N/A

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

N/A

/release-note-none

### CODE REVIEW

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go [hunk 1]
```diff
@@ -4586,6 +4590,15 @@ func (c *Cloud) nodeNameToProviderID(nodeName types.NodeName) (InstanceID, error
 	return KubernetesInstanceID(node.Spec.ProviderID).MapToAWSInstanceID()
 }
 
+// checkProtocol checks to make sure only TCP is supported for ELB
+func checkProtocol(proto v1.Protocol, annotations map[string]string) error {
+	if !isNLB(annotations) && proto != v1.ProtocolTCP {
```
Comment (bhagwat070919) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: How would this allow UDP protocol is this not the same as before. 

Does NLB check enough for creating the TCP and UDP ?

**Comment by jingweno** [comment]
@gnufied @micahhausler Could you review this? Thanks!

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go [hunk 2]
```diff
@@ -250,6 +250,9 @@ const (
 // The major consequence is that it is then not considered for AWS zone discovery for dynamic volume creation.
 var awsTagNameMasterRoles = sets.NewString("kubernetes.io/role/master", "k8s.io/role/master")
 
+// errorELBOnlySupportTCPLoadBalancer represents the error that only TCP LoadBalancer is supported for ELB
+var errorELBOnlySupportTCPLoadBalancer = fmt.Errorf("Only TCP LoadBalancer is supported for AWS ELB")
```
Comment (justinsb) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: No need to define a variable, unless you're exporting it

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go (hunk 1 above)
Comment (justinsb) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: We probably should affirmatively check the supported protocols for NLB; if a new protocol is introduced into k8s that NLB does not support, we probably want to catch it early

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go [hunk 3]
```diff
@@ -3521,9 +3524,10 @@ func (c *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, apiS
 
 	sslPorts := getPortSets(annotations[ServiceAnnotationLoadBalancerSSLPorts])
 	for _, port := range apiService.Spec.Ports {
-		if port.Protocol != v1.ProtocolTCP {
-			return nil, fmt.Errorf("Only TCP LoadBalancer is supported for AWS ELB")
+		if err := checkProtocol(port.Protocol, annotations); err != nil {
+			return nil, err
 		}
+
```
Comment (justinsb) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: I think you also need to think about HealthCheckProtocol, and also what about when when we are listening on port udp & tcp on the same port - I think we need to worry about the protocol + port now in the primary key of maps e.g. https://github.com/kubernetes/kubernetes/blob/c7f9dd0bafe2f3328148a85eed9c18322b9f308e/staging/src/k8s.io/legacy-cloud-providers/aws/aws_loadbalancer.go#L189

**Comment by justinsb** [comment]
Thanks for the PR.  I think we need to plumb this through in a few more places, but would be great to have this functionality!

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go (hunk 2 above)
Comment (jingweno) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: I'm reusing `errorELBOnlySupportTCPLoadBalancer` in test: https://github.com/kubernetes/kubernetes/pull/80394/files#diff-38fe63834837be4a3b79a6260b5591d1R1404.

**Comment by menshikh-iv** [comment]
@jingweno thanks for PR, we really need this one :rocket: 


@justinsb any plans/timelines for this feature? Can we help?

**Comment by xmudrii** [comment]
As this PR is supposed to fix #79523 which is in the 1.17 milestone, I'm adding this PR to the milestone too.

/milestone v1.17

**Comment by tylerohlsen** [comment]
Are community contributions welcome to help move this PR along?  This seems to have stalled a bit.

**Comment by bhagwat070919** [comment]
Hi All,

As we are not seeing any update from the contributor for review comments resolution and updating the PR with the code changes required.

What's the path forward? 

Should we consider another PR with the mentioned listener port and health check code changes?

While it should have been considered as mentioned here: https://aws.amazon.com/blogs/aws/new-udp-load-balancing-for-network-load-balancer/

It is clearly mentioned that we should have health checks defined in production systems. 

My question is: Can we merge this code so that we can add what is required ? or we can just rewrite same thing with new changes required?


@jingweno @justinsb @menshikh-iv @xmudrii @M00nF1sh @micahhausler

**Comment by kikisdeliveryservice** [comment]
Bug Triage Release Team checking in. Code freeze for the 1.17 release is on November 18. Is this still intended for 1.17? Thanks!

**Comment by oded-dd** [comment]
Can we get a reply on the intended date? It is super important for us.

**Comment by kikisdeliveryservice** [comment]
Hi all, 

Code freeze is coming up Nov 14th.  Is this staying in 1.17 or being bumped to 1.18?

**Comment by stefansedich** [comment]
Any push to get this one done? just ran into this myself @jingweno were you going to continue with this or did you need some help wrapping it up?

**Comment by jmatt-cavli** [comment]
Any confirmation on the fix for [#79523](https://github.com/kubernetes/kubernetes/issues/79523). Will this be available from 1.18?

**Comment by chancez** [comment]
This has tests, and is like 3 lines of code. Can we get this prioritized? It's been open for ~6 months.

**Comment by chancez** [comment]
ping @kubernetes/sig-cloud-provider @justinsb @zmerlynn @gnufied @jsafrane @micahhausler @mcrute @m00nf1sh

**Comment by chancez** [comment]
and @kubernetes/sig-cloud-provider-pr-reviews

**Comment by brutog** [comment]
This is important to us as well. We need to run a DNS application in kubernetes behind an NLB.

**Comment by gianarb** [comment]
Hello @justinsb  !
Bug Triage team here for the 1.18 release. This is a friendly reminder that code freeze is scheduled for 5 March. Is this issue still intended for milestone 1.18, or should it extend to the next one? Thanks in advance!

**Comment by andrewsykim** [comment]
Wondering if it's worthwhile to wait for https://github.com/kubernetes/enhancements/pull/1438 to be implemented before addressing this.

**Comment by gianarb** [comment]
Hello @andrewsykim thank for your feedback, do you think this will be merged soon? if https://github.com/kubernetes/enhancements/pull/1438 is required

**Comment by nckturner** [comment]
@andrewsykim I don't think they should conflict, it seems like #1438 would relax kubernetes service API validation to allow mixed protocols on ports.  A PR like this one would be required as a prerequisite for #1438 anyway, since it appears we currently don't allow UDP at all. 

What's missing in this PR? As @justinsb pointed out, the health check protocol is currently chosen from the non-UDP protocols, so we need to consider UDP when choosing health check port/protocol.

**Comment by gianarb** [comment]
Hello @justinsb  @kubernetes/sig-cloud-provider-bugs !
Bug Triage team here for the 1.18 release. This is a friendly reminder that code freeze is scheduled for 5 March. Is this issue still intended for milestone 1.18, or should it extend to the next one? Thanks in advance!

**Comment by trinitronx** [comment]
> What's missing in this PR? As @justinsb pointed out, the health check protocol is currently chosen from the non-UDP protocols, so we need to consider UDP when choosing health check port/protocol.

This is a known caveat when using a [UDP port listener on AWS NLBs][1].  The AWS blog post mentions this, and suggests adding an alternate TCP health check port (`80` for its' example) to workaround this issue.

> Health Checks – As I mentioned above, health checks must be done using TCP, HTTP, or HTTPS.

For the case of a UDP Kubernetes `Service`, one idea is perhaps implementing this as a multi-protocol `UDP + TCP` dynamic port (or dual `UDP + TCP` `NodePort`) such that the health check will function over `TCP`, while the service functions over `UDP`.  In some cases such as DNS (both port `53 UDP` & `53 TCP`), this may work without any Kubernetes-specific UDP to TCP health checking machinery.

For example, the blog post mentions this at the end (_TLDR;_):

> Multiple Protocols – A single Network Load Balancer can handle both TCP and UDP traffic. You can add another listener to an existing load balancer to gain UDP support, as long as you use distinct ports. In situations such as DNS where you need support for both TCP and UDP on the same port, you can set up a multi-protocol target group and a multi-protocol listener (use TCP_UDP for the listener type and the TargetGroup).


[1]: https://aws.amazon.com/blogs/aws/new-udp-load-balancing-for-network-load-balancer/

**Comment by smourapina** [comment]
We are now in code freeze for 1.18, so I will go ahead and set the milestone to 1.19. Thanks and please let me know if you have questions!
/milestone v1.19

**Comment by jmatt-cavli** [comment]
Hi All,

I am very sorry to say that, such a huge ecosystem like Kubernetes is
simply just forwarding such an important feature/requirement to next
milestones for last three releases, is really discouraging for the entire
community.

It’s purely my personal opinion.

**Comment by bai** [comment]
👋 Hello from Bug Triage team! This PR has not been updated for a while, so I'd like to check its status. The code freeze is starting June 25th (about 4 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the PR is tagged for 1.19, is it still planned for this release?

**Comment by bai** [comment]
👋 Hey @jingweno, a friendly bump here as code freeze starts tomorrow 🙏

**Comment by TBBle** [comment]
Superseded by #92109 which has now been merged.

**Review by andrewsykim — COMMENTED** [review]
/close

**Pull request closed without merging**